src/sentry/workflow_engine/processors/detector.py

```diff
@@ -0,0 +1,36 @@
+import logging
+
+from sentry.workflow_engine.models import Detector
+from sentry.workflow_engine.models.data_source import DataPacket
+from sentry.workflow_engine.models.detector import DetectorEvaluationResult
+
+logger = logging.getLogger(__name__)
+
+
+def process_detectors(
+    data_packet: DataPacket, detectors: list[Detector]
+) -> list[tuple[Detector, list[DetectorEvaluationResult]]]:
+    results = []
+    for detector in detectors:
+        handler = detector.detector_handler
+        if not handler:
+            continue
+        detector_results = handler.evaluate(data_packet)
+        detector_group_keys = set()
+        for result in detector_results:
+            if result.state_update_data:
+                if result.state_update_data.group_key in detector_group_keys:
+                    # This shouldn't happen - log an error and continue on, but we should investigate this.
+                    logger.error(
+                        "Duplicate detector state group keys found",
+                        extra={
+                            "detector_id": detector.id,
+                            "group_key": result.state_update_data.group_key,
+                        },
+                    )
+                detector_group_keys.add(result.state_update_data.group_key)
+
+        if detector_results:
+            results.append((detector, detector_results))
```

Comment:
🤔 i think process_workflows will also need the data_packet - are you thinking wherever this is invoked we rematch the `data_packet, list(DetectorEvaluationResult)` for `process_workflows`? (just thinking i'll add a `process_workflow_engine(data_packet: DataPacket)` method to simplify the common case if so)

Reply:
We only pass in a single packet into this function, so it's probably not too helpful to return the packet since it's implied by the call. I'm not strongly against it though if you'd prefer we do. I think you could just pair the results and pass them onto `process_workflows`, although I'm not sure where that call is happening

Reply:
nah, i think this is fine. i was just curious how you were thinking of pairing it up.

i like the idea of keeping these a bit more pure. i'll refactor process_data_sources a little with that in mind.